### ISSUE TITLE
kubelet CSRs: support out-of-tree attestation

### ISSUE THREAD

**Issue by ericchiang** [kubernetes/kubernetes#57164]
/kind feature

cc @mikedanese 
cc @kubernetes/sig-auth-feature-requests 
cc @kubernetes/sig-cluster-lifecycle-feature-requests 

TL;DR - How much of this do we actually want to do in-tree?

## Background

Today, kubelets present CSRs without any additional data. Approval process look at an individual CSR and generally say "this bootstrapping token is allowed to request a client cert, so I'll approve it." For most installs, bootstrapping tokens are allowed to request certificates for _any_ node. There's no recommended way for saying "this token is only valid for this exact node name" and our in-tree approvers isn't expressive enough to enforce this.

This is a blocker for kubelet serving certificate rotation https://github.com/kubernetes/features/issues/267. We want to be able to validate that we're issuing a serving cert to the right server.

We want to solve this problem through real attestation where a node actually presents some evidence that it's the node that's in the CSR CN and/or SANs. Potential kinds of attestation include:

* ACME challenges
  * Proposal by @mikedanese: https://docs.google.com/document/d/1l85qgbH-CoyCst7Ksuc12VyOr7rvDmhPJ5kmB1fDnS4/edit?ts=590a3fa5
* Cloud metadata APIs for validating instances
  * GCE: https://cloud.google.com/compute/docs/instances/verifying-instance-identity
  * AWS (this strategy is contentions): https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-identity-documents.html
* TPM values

There are no plugin points for the kubelet CSR flows today, though https://github.com/kubernetes/kubernetes/issues/55968 was intended to be a way of adding additional data to the CSR request.

On the approver side, we also want to be able to consult a machine catalog to ensure the node is actually in our cluster. "Is this instance in our auto-scaling group?"

## Open questions

As more installs start adopting the kubelet rotation, my question is:
1. Is https://github.com/kubernetes/kubernetes/issues/55968 + a custom authenticator a sufficient plugin mechanism for attestation?
2. Are there plans for other plugin mechanisms?
3. Do we just want to do this externally to the kubelet and have an out-of-tree node agent in charge of the CSR and attestation?

The more I think about this, the more I'm leaning towards option 3. Just externalize the whole thing instead of trying to design a plugin mechanism.

Would like to hear others' thoughts, because I'm a little unclear about where this feature is headed.
